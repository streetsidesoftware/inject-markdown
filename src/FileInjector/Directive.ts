import { parseRelativeUrl, type RelURL } from '../util/url_helper.js';

const injectDirectiveRegExp = /^[ \t]*<!--+\s*@@inject(?<type>|-start|-end|-code)[:\s]\s*(?<file>.*?)-+->$/;

export type DirectiveType = 'start' | 'end' | 'code';

export interface Directive {
    type: DirectiveType;
    file: RelURL | undefined;
}

export function parseDirective(html: string): Directive | undefined {
    const m = html.match(injectDirectiveRegExp);
    if (!m || !m.groups) return undefined;

    const filePath = m.groups['file'].trim();
    const file = (filePath && parseRelativeUrl(filePath)) || undefined;
    const isEnd = m.groups['type'] === '-end';
    const isCode = (!isEnd && !file?.pathname.toLowerCase().endsWith('.md')) || m.groups['type'] === '-code';

    const d: Directive = {
        type: isEnd ? 'end' : isCode ? 'code' : 'start',
        file,
    };
    return d;
}
