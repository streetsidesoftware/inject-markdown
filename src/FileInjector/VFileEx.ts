import type { Data as VFileData, MessageOptions as VFileMessageOptions } from 'vfile';
import { VFile } from 'vfile';

import type { BufferEncoding } from '../FileSystemAdapter/FileSystemAdapter.js';

export type MessageOptions = VFileMessageOptions['place'];

export interface FileData extends VFileData {
    encoding: BufferEncoding;
    fileUrl: URL;
    cwdUrl?: URL;
    hasInjections?: boolean;
}

export class VFileEx extends VFile {
    readonly fileUrl: URL;
    data: FileData;

    constructor(value: string, data: FileData) {
        super({ path: data.fileUrl.pathname, value, data });
        this.fileUrl = data.fileUrl;
        this.data = data;
    }

    error(reason: string, place?: MessageOptions): ReturnType<VFile['message']> {
        const msg = this.message(reason, place);
        msg.fatal = true;
        return msg;
    }
}

export function isVFileEx(file: VFile | VFileEx): file is VFileEx {
    return file instanceof VFileEx;
}
