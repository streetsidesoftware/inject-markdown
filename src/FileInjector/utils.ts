import type { BufferEncoding } from '../FileSystemAdapter/FileSystemAdapter.js';

export function toString(content: string | Buffer | Uint8Array, encoding: BufferEncoding): string {
    if (typeof content === 'string') return content;

    if (content instanceof Buffer) {
        return content.toString(encoding);
    }

    return Buffer.from(content).toString(encoding);
}

export function toError(e: unknown): Error {
    if (e && typeof e === 'object') return e as Error;
    if (typeof e === 'string') return new Error(e);
    return new Error('Unknown');
}
