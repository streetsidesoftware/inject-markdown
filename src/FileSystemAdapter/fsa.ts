import * as fs from 'fs/promises';
import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';
import { isURL } from '../util/url_helper.js';
import fetch from 'node-fetch';

async function readFile(file: PathLike, encoding: BufferEncoding): Promise<string> {
    if (isURL(file)) {
        if (file.protocol === 'file:') {
            return await fs.readFile(file, encoding);
        }
        return await fetchUrl(file);
    }
    return await fs.readFile(file, encoding);
}

export function nodeFsa(): FileSystemAdapter {
    const fsa: FileSystemAdapter = {
        readFile,
        mkdir: fs.mkdir,
        writeFile: fs.writeFile,
    };

    return fsa;
}

async function fetchUrl(url: URL): Promise<string> {
    const response = await fetch(mapUrl(url));
    return await response.text();
}

function mapUrl(url: URL): string {
    let href = url.href;

    // Remap GitHub blobs
    href = href.replace(/^https?:\/\/github.com\/([^/]+\/[^/]+)\/blob\//, 'https://raw.githubusercontent.com/$1/');

    return href;
}
