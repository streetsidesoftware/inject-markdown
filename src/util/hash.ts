import { RelURL } from './url_helper.js';

export interface InjectInfo {
    heading?: string | undefined;
    /** code bock language */
    lang?: string | undefined;
    /** lines to use */
    lines?: [number, number] | undefined;
    tags?: string[] | undefined;
    params?: Map<string, string | string[]>;
}

export function parseHash(url: URL | RelURL): InjectInfo {
    const info = parseHashString(url.hash);

    if (!info.heading && !info.lines && info.tags?.length) {
        info.heading = info.tags[0];
    }

    return info;
}

const regExLineNumExpression = /^(L\d+)(?:-(L\d+))?$/;

export function parseHashString(hash: string): InjectInfo {
    hash = hash.replace(/^#/, '');
    if (!hash) return {};

    const info: InjectInfo = {};

    const params = new URLSearchParams(hash);
    const tags: string[] = [];

    const p = new Map<string, string | string[]>();

    function addParam(key: string, value: string) {
        const v = p.get(key);
        if (v !== undefined) {
            const isArray = Array.isArray(v);
            const a = isArray ? v : [v];
            a.push(value);
            !isArray && p.set(key, a);
            return;
        }
        p.set(key, value);
    }

    for (const [key, value] of params.entries()) {
        addParam(key, value);
        switch (key) {
            case 'code':
            case 'lang':
                info.lang = value;
                continue;
            case 'heading':
                info.heading = value;
                continue;
            case 'lines':
            case 'line':
                {
                    const range = parseLineNumbers(value);
                    if (range) {
                        info.lines = range;
                        continue;
                    }
                }
                break;
        }

        const lineRange = parseLineNumbers(key);
        if (lineRange) {
            info.lines = lineRange;
            continue;
        }

        !value && tags.push(key);
    }

    if (tags.length) {
        info.tags = tags;
    }

    if (p.size) {
        info.params = p;
    }

    return info;
}

function parseLineNumbers(ref: string): [number, number] | undefined {
    const match = ref.match(regExLineNumExpression);
    if (!match) return undefined;

    const start = match[1];
    const end = match[2] || start;
    const startNum = parseInt(start.slice(1));
    const endNum = parseInt(end.slice(1));
    if (startNum && endNum) {
        return [startNum, endNum];
    }
    return undefined;
}
