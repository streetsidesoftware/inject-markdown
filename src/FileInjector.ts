import assert from 'assert';
import type { Content, HTML, Root, Parent } from 'mdast';
import * as path from 'node:path';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import { unified } from 'unified';
import { is } from 'unist-util-is';
import { remove } from 'unist-util-remove';
import { visit } from 'unist-util-visit';
import { VFile } from 'vfile';
import { reporter } from 'vfile-reporter';
import { parentPort } from 'worker_threads';

import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';

type Node = Root | Content;

// const injectStartRegExp = /^[ \t]*<!---?\s*@@inject(?:-start)?:/gm;
// const injectEndRegExp = /^[ \t]*<!---?\s*@@inject-end:/gm;
// const injectDisable = /@@inject-disable\b/g;
// const injectEnable = /@@inject-enable\b/g;

const injectDirectiveRegExp = /^[ \t]*<!---?\s*@@inject(?<type>|-start|-end):\s*(?<file>.*?)-?-->$/;

// const commentStart = '<!---';
// const commentEnd = '-->';

const directiveRegExp = /^[ \t]*<!---?\s*@@inject\b/;

const directivePrefix = '@@inject';

export interface FileInjectorOptions {
    /** optional output directory */
    outputDir?: string;
    /** Current working directory */
    cwd?: string;
    /** Only clean the file, do not inject */
    cleanOnly?: boolean;
}

export class FileInjector {
    constructor(readonly fs: FileSystemAdapter, readonly options: FileInjectorOptions) {}

    /**
     * Process all injections for a file.
     * @param filePath - path to file to process
     * @param encoding - file encoding
     * @returns true if changed, otherwise false.
     */
    async processFile(filePath: PathLike, encoding: BufferEncoding = 'utf8'): Promise<boolean> {
        const file = await readFile(this.fs, filePath, encoding);
        return processFileInjections(file, this.fs, this.options);
    }
}

function processFileInjections(file: VFile, fs: FileSystemAdapter, options: FileInjectorOptions): Promise<boolean> {
    return __processFile();

    async function __processFile(): Promise<boolean> {
        const fileValue = file.value;
        const content = extractContent(file);
        const lineEnding = detectLineEnding(content);
        const result = await processFileContent();
        if (result.value === fileValue) {
            options.outputDir && (await writeResult(result));
            return false;
        }
        console.log(reporter(result));
        const resultAsString = extractContent(result);
        const resultContent = fixContentLineEndings(resultAsString, lineEnding, hasEofNewLine(content));
        if (content === resultContent) {
            options.outputDir && (await writeResult(result));
            return false;
        }
        await writeResult(result);
        return true;
    }

    async function writeResult(file: VFile): Promise<void> {
        const content = extractContent(file);
        const filePath = determineTargetPath(file);
        const dir = path.dirname(filePath);
        await fs.mkdir(dir, { recursive: true });
        return fs.writeFile(filePath, content, getEncoding(file));
    }

    function determineTargetPath(file: VFile): string {
        const outDir = options.outputDir;
        if (!outDir) return file.path;

        const cwd = path.resolve(options.cwd || file.cwd);
        const srcPath = path.resolve(cwd, file.path);
        const relPath = path.relative(cwd, srcPath);

        return path.join(path.resolve(outDir), relPath);
    }

    async function processFileContent(): Promise<VFile> {
        if (!extractContent(file).includes(directivePrefix)) {
            return file;
        }
        const result = await unified().use(remarkParse).use(processInjections).use(remarkStringify).process(file);
        return result;
    }

    function processInjections() {
        return async (root: Root, file: VFile): Promise<Root> => {
            root = deleteInjectedContent(root, file);
            if (options.cleanOnly) return root;
            root = await injectFiles(root);
            return root;
        };
    }

    async function injectFiles(root: Root): Promise<Root> {
        const directiveNodes = collectInjectionNodes(root);

        for (const node of directiveNodes) {
            if (!node.file) continue;
            await injectFile(node);
        }

        return root;
    }

    async function injectFile(directive: DirectiveNode): Promise<void> {
        if (!directive.file || directive.type !== 'start') return;
        const [encodedName, header] = directive.file.split('#', 2);
        const fileName = decodeURI(encodedName);
        console.log(fileName);

        const parent = directive.parent;
        const index = parent.children.indexOf(directive.node);
        assert(index >= 0);
        const addStart = directive.node.value.includes('@@inject-start');
        const root = await readAndParseMarkdownFile(fileName, header);
        remove(root, (n: Node) => isHtmlNode(n) && n.value.includes('@@inject'));
        const encodedFile = encodeURI(fileName) + (header ? '#' + encodeURIComponent(header) : '');
        const start: HTML = {
            type: 'html',
            value: `<!--- @@inject${addStart ? '-start' : ''}: ${encodedFile} --->`,
        };
        const end: HTML = {
            type: 'html',
            value: `<!--- @@inject-end: ${encodedFile} --->`,
        };
        parent.children.splice(index, 1, start, ...root.children, end);
    }

    async function readAndParseMarkdownFile(fileName: string, _atHeader: string | undefined): Promise<Root> {
        const dir = file.dirname || process.cwd();
        const resolvedFile = path.resolve(dir, fileName);
        try {
            const vFile = await readFile(fs, resolvedFile);
            return parseMarkdownFile(vFile);
        } catch (e) {
            file.message(e as Error);
            return unified().use(remarkParse).parse(`\
<!---
  Failed to read "${fileName}"
--->`);
        }
    }

    function parseMarkdownFile(file: VFile): Root {
        return unified().use(remarkParse).parse(file);
    }
}

function deleteInjectedContent(root: Root, file: VFile): Root {
    const directiveNodes = collectInjectionNodes(root);
    const pairs = findInjectionPairs(directiveNodes, file);

    interface BaseNode {
        type: string;
    }

    const starts = new Set<BaseNode>(
        pairs
            .map((p) => p.end && p.start)
            .filter(isDefined)
            .map((d) => d.node)
    );
    const ends = new Set<BaseNode>(
        pairs
            .map((p) => p.start && p.end)
            .filter(isDefined)
            .map((d) => d.node)
    );
    const toDelete = new Set<BaseNode>(
        pairs
            .map((p) => p.end)
            .filter(isDefined)
            .map((d) => d.node)
    );

    let deleteDepth = 0;

    visit(
        root,
        () => true,
        (node: BaseNode) => {
            if (deleteDepth) {
                toDelete.add(node);
            }
            if (starts.has(node)) {
                ++deleteDepth;
            }
            if (ends.has(node)) {
                --deleteDepth;
                assert(deleteDepth >= 0);
            }
        }
    );

    remove(root, (node: BaseNode) => toDelete.has(node));

    return root;
}

interface DirectivePair {
    start?: DirectiveNode | undefined;
    end?: DirectiveNode | undefined;
}

interface Directive {
    type: 'start' | 'end';
    file: string;
}

interface DirectiveNode extends Partial<Directive> {
    node: HTML;
    parent: Parent;
}

function findInjectionPairs(nodes: DirectiveNode[], vfile: VFile): DirectivePair[] {
    function validate(n: DirectiveNode): n is Required<DirectiveNode> {
        if (!n.type) {
            vfile.message('Unable to parse @@inject directive.', n.node.position);
            return false;
        }
        if (n.type === 'start' && !n.file) {
            vfile.message('Missing injection filename.', n.node.position);
            return false;
        }
        return true;
    }

    const dnp = nodes.filter(validate);

    const pairs: DirectivePair[] = [];

    let last: DirectiveNode | undefined = undefined;
    for (const n of dnp.reverse()) {
        if (n.type === 'start') {
            if (last?.file && last?.file !== n.file) {
                vfile.message(`Unmatched @@inject-end "${last.file || ''}"`, last.node.position);
                pairs.push({ end: last });
                last = undefined;
            }
            pairs.push({ start: n, end: last });
            last = undefined;
            continue;
        }
        if (last) {
            vfile.message(`Unmatched @@inject-end "${last.file || ''}"`, last.node.position);
            pairs.push({ end: last });
        }
        last = n;
    }

    return pairs.reverse();
}

function parseDirectiveNode(node: HTML): Directive | undefined {
    return parseDirective(node.value);
}

function parseDirective(html: string): Directive | undefined {
    const m = html.match(injectDirectiveRegExp);
    if (!m || !m.groups) return undefined;

    const d: Directive = {
        type: m.groups['type'] === '-end' ? 'end' : 'start',
        file: m.groups['file'].trim(),
    };
    return d;
}

function collectInjectionNodes(root: Root): DirectiveNode[] {
    const nodes: DirectiveNode[] = [];
    visit(root, isInjectNode, (node, _index, parent) => {
        parent && nodes.push({ node, parent });
    });

    const dNodes = nodes.map((node) => ({ ...node, ...(parseDirectiveNode(node.node) || {}) }));
    return dNodes;
}

interface VFileData {
    encoding?: BufferEncoding;
}

async function readFile(fs: FileSystemAdapter, path: PathLike, encoding: BufferEncoding = 'utf8'): Promise<VFile> {
    const value = await fs.readFile(path, encoding);
    return new VFile({ path, value, data: { encoding } });
}

function detectLineEnding(content: string): string {
    const pos = content.indexOf('\n');
    return content[pos - 1] === '\r' ? '\r\n' : '\n';
}

function isHtmlNode(n: Node | unknown): n is HTML {
    return is(n, 'html');
}

function isInjectNode(n: Node | unknown): n is HTML {
    if (!isHtmlNode(n)) {
        return false;
    }
    return directiveRegExp.test(n.value);
}

function extractContent(file: VFile, defaultEncoding?: BufferEncoding): string {
    return toString(file.value, getEncoding(file, defaultEncoding));
}

function getEncoding(file: VFile, defaultEncoding: BufferEncoding = 'utf8'): BufferEncoding {
    const data: VFileData = file.data;
    return data.encoding || defaultEncoding;
}

function toString(content: string | Buffer, encoding: BufferEncoding): string {
    return typeof content === 'string' ? content : content.toString(encoding);
}

function fixContentLineEndings(content: string, lineEnding: string, fixEofNewLine: boolean): string {
    const fixed = content.replace(/\r?\n/g, lineEnding);
    return fixEofNewLine && !hasEofNewLine(fixed) ? fixed + lineEnding : fixed;
}

function hasEofNewLine(content: string): boolean {
    return content[content.length - 1] === '\n';
}

function isDefined<T>(v: T | undefined | null): v is T {
    return v !== undefined && v !== null;
}
