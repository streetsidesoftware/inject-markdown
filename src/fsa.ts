import * as fs from 'fs/promises';
import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';
import { isURL } from './url_helper.js';
import fetch from 'node-fetch';

async function readFile(file: PathLike, encoding: BufferEncoding): Promise<string> {
    if (isURL(file)) {
        if (file.protocol === 'file:') {
            return await fs.readFile(file, encoding);
        }
        const response = await fetch(file.href);
        return await response.text();
    }
    return await fs.readFile(file, encoding);
}

export function nodeFsa(): FileSystemAdapter {
    const fsa: FileSystemAdapter = {
        readFile,
        mkdir: fs.mkdir,
        writeFile: fs.writeFile,
    };

    return fsa;
}
