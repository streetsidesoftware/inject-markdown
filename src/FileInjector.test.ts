import { fileURLToPath } from 'node:url';
import * as path from 'path';
import { format } from 'util';
import { describe, expect, MockedFunction, test, vi } from 'vitest';
import { FileInjector, Logger } from './FileInjector.js';
import type { FileSystemAdapter, PathLike, BufferEncoding } from './FileSystemAdapter.js';
import { nodeFsa } from './fsa.js';

const __file__ = fileURLToPath(import.meta.url);
const __dirname__ = path.dirname(__file__);
const __root__ = path.join(__dirname__, '..');

const appFsa = nodeFsa();

describe('FileInjector', () => {
    test('FileInjector', () => {
        const fi = new FileInjector(createFSA(), {});
        expect(fi).toBeDefined();
    });

    test.each`
        file           | options                              | expectedResult
        ${'README.md'} | ${{ cwd: 'fixtures/no-injections' }} | ${false}
    `('processFile no change', async ({ file, options, expectedResult }) => {
        options.cwd = options.cwd || __root__;
        const fsa = createFSA();
        const fi = new FileInjector(fsa, options);
        const r = await fi.processFile(file);
        expect(r).toBe(expectedResult);
        expect(fsa.mkdir).not.toBeCalled();
    });

    test.each`
        file                                 | options                                                  | expectedResult | expectedFile
        ${'fixtures/vacations/vacations.md'} | ${{}}                                                    | ${true}        | ${'fixtures/vacations/vacations.md'}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_' }}                                | ${true}        | ${'_out_/fixtures/vacations/vacations.md'}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_', verbose: true }}                 | ${true}        | ${'_out_/fixtures/vacations/vacations.md'}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_', silent: true }}                  | ${true}        | ${'_out_/fixtures/vacations/vacations.md'}
        ${'vacations.md'}                    | ${{ cwd: 'fixtures/vacations/', outputDir: '_out_' }}    | ${true}        | ${'_out_/vacations.md'}
        ${'README.md'}                       | ${{ cwd: 'fixtures/no-injections', outputDir: '_out_' }} | ${false}       | ${'_out_/README.md'}
    `('processFile $file $options', async ({ file, options, expectedResult, expectedFile }) => {
        const logger = createLogger();
        options.cwd = options.cwd || __root__;
        options.color = options.color ?? false;
        options.logger = logger;
        expectedFile = path.resolve(__root__, expectedFile);
        const fsa = createFSA();
        const fi = new FileInjector(fsa, options);
        const r = await fi.processFile(file);
        expect(r).toBe(expectedResult);
        expect(fsa.mkdir).toHaveBeenCalledWith(path.dirname(expectedFile), { recursive: true });
        expect(normalizeWriteFileCalls(fsa.writeFile)).toMatchSnapshot();
        expect(logger.history).toMatchSnapshot();
    });
});

function normalizeWriteFileCalls(
    writeFile: MockedFileSystemAdapter['writeFile']
): MockedFileSystemAdapter['writeFile']['mock']['calls'] {
    const calls = writeFile.mock.calls;
    const normalized = calls.map(
        ([pathLike, data, encoding]) =>
            [normalizePath(pathLike), data, encoding] as Parameters<FileSystemAdapter['writeFile']>
    );
    return normalized;
}

function normalizePath(pathLike: PathLike): PathLike {
    if (typeof pathLike !== 'string') return pathLike;

    const p = path.relative(__root__, pathLike);
    return p.replace(/\\/g, '/');
}

type MockedFileSystemAdapter<T extends FileSystemAdapter = FileSystemAdapter> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [K in keyof T]: T[K] extends (...args: any) => any ? MockedFunction<T[K]> : T[K];
};

interface Store {
    get(file: PathLike): string | undefined;
    set(file: PathLike, data: string | undefined): void;
}

function createStore(): Store {
    const store = new Map<string, string>();

    function get(file: PathLike) {
        const p = normalizePath(file).toString();
        return store.get(p);
    }

    function set(file: PathLike, data: string | undefined) {
        const p = normalizePath(file).toString();
        if (typeof data !== 'string') {
            store.delete(p);
            return;
        }
        store.set(p, data);
    }

    return { get, set };
}

interface FSA extends MockedFileSystemAdapter {
    store: Store;
}

function createFSA(): FSA {
    const store = createStore();

    async function readFile(p: PathLike, e: BufferEncoding): Promise<string> {
        const found = store.get(p);
        if (typeof found === 'string') return found;
        const data = await appFsa.readFile(p, e);
        store.set(p, data);
        return data;
    }

    async function writeFile(file: PathLike, data: string, _encoding: string) {
        store.set(file, data);
    }

    const fsaMethods: FileSystemAdapter = {
        readFile: vi.fn().mockImplementation(readFile),
        writeFile: vi.fn().mockImplementation(writeFile),
        mkdir: vi.fn().mockImplementation(async (_path: PathLike) => undefined),
    };

    const ma = fsaMethods as MockedFileSystemAdapter;

    const fsa: FSA = {
        ...ma,
        store,
    };

    return fsa;
}

function createLogger() {
    type Target = 'error' | 'log' | 'warn' | 'stdout' | 'stderr';
    const history: { target: Target; text: string }[] = [];

    function f(target: Target): typeof console.log {
        function log(...params: Parameters<typeof console.log>) {
            history.push({ target, text: format(...params) });
        }
        return log;
    }
    const log = f('log');
    const error = f('error');
    const warn = f('warn');
    const stderr = f('stderr');
    const stdout = f('stdout');

    const logger = {
        log: vi.fn().mockImplementation(log),
        error: vi.fn(error),
        warn: vi.fn(warn),
        writeStdout: vi.fn((t) => stdout(t)),
        writeStderr: vi.fn((t) => stderr(t)),
        history,
    };
    return logger satisfies Logger;
}
