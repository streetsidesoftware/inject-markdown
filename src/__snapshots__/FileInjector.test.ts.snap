// Vitest Snapshot v1

exports[`FileInjector > processFile 'README.md' { cwd: 'fixtures/no-injections', …(1) } 1`] = `
[
  [
    "_out_/README.md",
    "# Without injections

This file does not have any injections.
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'README.md' { cwd: 'fixtures/no-injections', …(1) } 2`] = `
[
  {
    "target": "stderr",
    "text": "README.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_' } 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_' } 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', silent: true } 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', silent: true } 2`] = `[]`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', verbose: true } 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', verbose: true } 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": "
  parts/destinations.md",
  },
  {
    "target": "stderr",
    "text": "
  parts/prices.md",
  },
  {
    "target": "stderr",
    "text": "
  parts/destinations.md#Amsterdam",
  },
  {
    "target": "stderr",
    "text": "
  parts/code.md",
  },
  {
    "target": "stderr",
    "text": "
  done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' {} 1`] = `
[
  [
    "fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' {} 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'vacations.md' { cwd: 'fixtures/vacations/', …(1) } 1`] = `
[
  [
    "_out_/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'vacations.md' { cwd: 'fixtures/vacations/', …(1) } 2`] = `
[
  {
    "target": "stderr",
    "text": "vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;
